File: src/main/java/dev/obscuria/obscure_tooltips/client/tooltip/TooltipStyle.java
package dev.obscuria.obscure_tooltips.client.tooltip;

import com.github.bsideup.jabel.Desugar;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import dev.obscuria.obscure_tooltips.client.registry.ResourceRegistry;
import dev.obscuria.obscure_tooltips.client.registry.TooltipRegistries;
import dev.obscuria.obscure_tooltips.client.tooltip.element.effect.TooltipEffect;
import dev.obscuria.obscure_tooltips.client.tooltip.element.frame.TooltipFrame;
import dev.obscuria.obscure_tooltips.client.tooltip.element.icon.TooltipIcon;
import dev.obscuria.obscure_tooltips.client.tooltip.element.panel.TooltipPanel;
import dev.obscuria.obscure_tooltips.client.tooltip.element.slot.TooltipSlot;
import net.minecraft.util.JsonUtils;
import net.minecraft.util.ResourceLocation;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

@Desugar
public record TooltipStyle(
        Optional<TooltipPanel> panel,
        Optional<TooltipFrame> frame,
        Optional<TooltipSlot> slot,
        Optional<TooltipIcon> icon,
        List<TooltipEffect> effects) {

    public static final TooltipStyle EMPTY = new TooltipStyle(
            Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Collections.emptyList());

    public static TooltipStyle fromJson(JsonObject json) {
        return new TooltipStyle(
                resolve(json, "panel", TooltipRegistries.TOOLTIP_PANEL),
                resolve(json, "frame", TooltipRegistries.TOOLTIP_FRAME),
                resolve(json, "slot", TooltipRegistries.TOOLTIP_SLOT),
                resolve(json, "icon", TooltipRegistries.TOOLTIP_ICON),
                resolveEffects(json));
    }

    public TooltipStyle merge(TooltipStyle other) {
        return new TooltipStyle(
                panel.isPresent() ? panel : other.panel,
                frame.isPresent() ? frame : other.frame,
                slot.isPresent() ? slot : other.slot,
                icon.isPresent() ? icon : other.icon,
                mergeEffects(other.effects));
    }

    private List<TooltipEffect> mergeEffects(List<TooltipEffect> other) {
        if (effects.isEmpty()) return other;
        if (other.isEmpty()) return effects;

        final var result = new ArrayList<>(effects);
        for (var effect : other) {
            if (!effect.canApply(result)) continue;
            result.add(effect);
        }

        return result;
    }

    private static <T> Optional<T> resolve(JsonObject json, String key, ResourceRegistry<T> registry) {
        if (!json.has(key)) {
            return Optional.empty();
        }
        return Optional.ofNullable(registry.get(new ResourceLocation(JsonUtils.getString(json, key))));
    }

    private static List<TooltipEffect> resolveEffects(JsonObject json) {
        if (!json.has("effects")) {
            return Collections.emptyList();
        }
        final List<TooltipEffect> effects = new ArrayList<>();
        for (JsonElement element : JsonUtils.getJsonArray(json, "effects")) {
            final TooltipEffect effect = TooltipRegistries.TOOLTIP_EFFECT.get(new ResourceLocation(element.getAsString()));
            if (effect != null) {
                effects.add(effect);
            }
        }
        return effects;
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/TooltipStyle.kt
package dev.obscuria.tooltips.client

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.obscuria.tooltips.client.element.effect.ITooltipEffect
import dev.obscuria.tooltips.client.element.frame.ITooltipFrame
import dev.obscuria.tooltips.client.element.icon.ITooltipIcon
import dev.obscuria.tooltips.client.element.panel.ITooltipPanel
import dev.obscuria.tooltips.client.element.slot.ITooltipSlot
import dev.obscuria.tooltips.registry.TooltipsRegistries
import java.util.*

data class TooltipStyle(
    val panel: Optional<ITooltipPanel> = Optional.empty(),
    val frame: Optional<ITooltipFrame> = Optional.empty(),
    val slot: Optional<ITooltipSlot> = Optional.empty(),
    val icon: Optional<ITooltipIcon> = Optional.empty(),
    val effects: List<ITooltipEffect> = emptyList()
) {

    fun merge(other: TooltipStyle): TooltipStyle = TooltipStyle(
        panel = panel.or(other::panel),
        frame = frame.or(other::frame),
        slot = slot.or(other::slot),
        icon = icon.or(other::icon),
        effects = effects.merge(other.effects)
    )

    companion object {

        val EMPTY: TooltipStyle = TooltipStyle()
        val CODEC: Codec<TooltipStyle> = TooltipsRegistries.Resource.TOOLTIP_STYLE.byNameCodec()
        val DIRECT_CODEC: Codec<TooltipStyle> = RecordCodecBuilder.create { codec ->
            codec.group(
                ITooltipPanel.CODEC
                    .optionalFieldOf("panel")
                    .forGetter(TooltipStyle::panel),
                ITooltipFrame.CODEC
                    .optionalFieldOf("frame")
                    .forGetter(TooltipStyle::frame),
                ITooltipSlot.CODEC
                    .optionalFieldOf("slot")
                    .forGetter(TooltipStyle::slot),
                ITooltipIcon.CODEC
                    .optionalFieldOf("icon")
                    .forGetter(TooltipStyle::icon),
                ITooltipEffect.CODEC.listOf()
                    .fieldOf("effects")
                    .forGetter(TooltipStyle::effects)
            ).apply(codec, ::TooltipStyle)
        }

        private fun List<ITooltipEffect>.merge(other: List<ITooltipEffect>): List<ITooltipEffect> {

            if (this.isEmpty()) return other
            if (other.isEmpty()) return this

            val result = this.toMutableList()
            for (effect in other) {
                if (!effect.canApply(result)) continue
                result.add(effect)
            }
            return result
        }
    }
}